tools/discovery/discovery.go

```diff
@@ -0,0 +1,131 @@
+package discovery
+
+import (
+	"bytes"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+	"github.com/xiekeyang/oci-discovery/tools/object"
+)
+
+var (
+	defaultTrans = &http.Transport{
+		TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+	}
+)
+
+func DiscoveryHandler(context *cli.Context) error {
+	var (
+		name  = context.Args()[0]
+		roots []v1.Descriptor
+	)
+
+	v, err := paramsParser(name)
+	if err != nil {
+		return err
+	}
+
+	engines, err := refEnginesFetching(v)
+	if err != nil {
+		return err
+	}
+
+	for _, engine := range engines.RefEngines {
+		var ur urlResolver = urlResolver(engine.Uri)
```

Comment:
Ref-engine setup may not be free (e.g. if a ref-engine decides to connect to a service as part of initialization).  I recommend having the ref-engine-discovery call (`refEnginesFetching` in 4829c1e7) return configuration `[]map[string]interface{}` and then only instantiate the ref-engine objects as needed inside this loop.